mars/services/task/execution/ray/executor.py

```diff
@@ -118,29 +124,66 @@ def execute_subtask(
     subtask_id: str,
     subtask_chunk_graph: ChunkGraph,
     output_meta_keys: Set[str],
-    input_keys: List[str],
     *inputs,
 ):
-    logger.info("Begin to execute subtask: %s", subtask_id)
+    """The function used for execute subtask in ray task.
+    If subtask is shuffle mapper, no chunk meta will be returned, otherwise return chunk mata."""
     ensure_coverage()
     subtask_chunk_graph = deserialize(*subtask_chunk_graph)
-    # inputs = [i[1] for i in inputs]
+    subtask_digraph = subtask_chunk_graph.to_dot()
+    logger.info("Begin to execute subtask: %s", subtask_id)
+    # optimize chunk graph.
+    subtask_chunk_graph = _optimize_subtask_graph(subtask_chunk_graph)
+    start_chunks = list(subtask_chunk_graph.iter_indep())
+    maybe_mapper_chunk = subtask_chunk_graph.result_chunks[0]
+    is_mapper = (
+        isinstance(maybe_mapper_chunk.op, MapReduceOperand)
+        and maybe_mapper_chunk.op.stage == OperandStage.map
+    )
+    if isinstance(start_chunks[0].op, FetchShuffle):
+        assert len(start_chunks) == 1, start_chunks
+        # the subtask is a reducer subtask
+        n_mappers = len(inputs)
+        # some reducer may have multiple output chunks, see `PSRSshuffle._execute_reduce` and
+        # https://user-images.githubusercontent.com/12445254/168569524-f09e42a7-653a-4102-bdf0-cc1631b3168d.png
+        reducer_chunks = subtask_chunk_graph.successors(start_chunks[0])
+        reducer_operands = set(c.op for c in reducer_chunks)
+        assert len(reducer_operands) == 1, (
+            reducer_operands,
+            reducer_chunks,
+            subtask_digraph,
+        )
+        reducer_operand = reducer_chunks[0].op
+        reducer_index = reducer_operand.reducer_index
+        # mock input keys, keep this in sync with `MapReducerOperand#_iter_mapper_key_idx_pairs`
+        input_keys = [(i, reducer_index) for i in range(n_mappers)]
+    else:
+        input_keys = [c.key for c in start_chunks if isinstance(c.op, Fetch)]
```

Comment:
The `iter_indep` order is guaranteed, please transfer the `input_keys` for non shuffle subtasks.

Reply:
I omitted `input_keys ` on purpose, so the meta for task spec will be small nad serialization cost is small too.

Reply:
My point is the order of `iter_indep` is not guaranteed through serialize / deserialze. If the `input_keys` is omitted here, then we should add more checks to keep the order when serializing.

Reply:
How about add `sort_by` parameter to `iter_indep` oor add a `iter_indep_sorted` API to `DirectedGraph`. I did a rough tests, sort chunks is much faster tghen serialization. And if we omit `input_keys`, the task spec is small, we can also cache more task spec for lineage reconstruction 
![image](https://user-images.githubusercontent.com/12445254/174006854-e2d2011d-e624-4eea-89c2-3683f479a8f1.png)

Reply:
It depends on the keys size, the sort algorithm has `O(n)log(n)` time complexity. If the keys are small, sort is much faster than serialization.
![image](https://user-images.githubusercontent.com/12445254/174758349-991459ef-de8d-4c16-a2a1-ee4ee4efd48c.png)
And the input keys in subtask chunk graph is less than 5 normally so the cost is ignorable.

I agree with you that sort should be used sparely. If we can avoid sort, the code will be more clean. If we serializing dict by two list, the order will be fixed. But sort is more explict and customizing may should be put in another PR if we really need it?